src/System.Private.CoreLib/shared/System/Number.NumberToDouble.cs

```diff
@@ -0,0 +1,450 @@
+            }
+
+            return val;
+        }
+
+        // slower high precision version of Mul64 for computation of the tables
+        private static ulong Mul64Precise(ulong a, ulong b, ref int exp)
+        {
+            ulong hilo = ((Mul32x32To64((uint)(a >> 32), (uint)(b)) >> 1)
+                       + (Mul32x32To64((uint)(a), (uint)(b >> 32)) >> 1)
+                       + (Mul32x32To64((uint)(a), (uint)(b)) >> 33)) >> 30;
+
+            ulong val = Mul32x32To64((uint)(a >> 32), (uint)(b >> 32))
+                      + (hilo >> 1)
+                      + (hilo & 1);
+
+            // normalize
+            if ((val & 0x8000000000000000) == 0)
+            {
+                val <<= 1;
+                exp--;
+            }
+
+            return val;
+        }
+
+        private static double NumberToDouble(ref NumberBuffer number)
+        {
+#if DEBUG
+            
+            if (!s_CheckedTables)
+            {
+                CheckTables();
+                s_CheckedTables = true;
+            }
+#endif
+
+            char* src = number.digits;
+            int total = GetLength(src);
+            int remaining = total;
+
+            // skip the leading zeros
+            while (src[0] == '0')
+            {
+                remaining--;
+                src++;
+            }
+
+            if (remaining == 0)
+            {
+                return number.sign ? -0.0 : 0.0;
+            }
+
+            int count = Math.Min(remaining, 9);
+            remaining -= count;
+            ulong val = DigitsToInt(src, count);
+
+            if (remaining > 0)
+            {
+                count = Math.Min(remaining, 9);
+                remaining -= count;
+
+                // get the denormalized power of 10
+                uint mult = (uint)(rgval64Power10[count - 1] >> (64 - rgexp64Power10[count - 1]));
+                val = Mul32x32To64((uint)(val), mult) + DigitsToInt(src + 9, count);
+            }
+
+            int scale = number.scale - (total - remaining);
+            int absscale = Math.Abs(scale);
+            if (absscale >= 22 * 16)
+            {
+                // overflow / underflow
+                if (scale > 0)
+                {
+                    return number.sign ? double.NegativeInfinity : double.PositiveInfinity;
+                }
+                else
+                {
+                    return number.sign ? -0.0 : 0.0;
+                }
+            }
+
+            int exp = 64;
+
+            // normalize the mantissa
+            if ((val & 0xFFFFFFFF00000000) == 0) { val <<= 32; exp -= 32; }
+            if ((val & 0xFFFF000000000000) == 0) { val <<= 16; exp -= 16; }
+            if ((val & 0xFF00000000000000) == 0) { val <<= 8; exp -= 8; }
+            if ((val & 0xF000000000000000) == 0) { val <<= 4; exp -= 4; }
+            if ((val & 0xC000000000000000) == 0) { val <<= 2; exp -= 2; }
+            if ((val & 0x8000000000000000) == 0) { val <<= 1; exp -= 1; }
```

Comment:
Are some readability underscores in these various numbers  in this file worthwhile?

Reply:
Possibly, but I would prefer to fix that with a new algorithm that actually fixes the parsing code to be correct.

Reply:
Fixed, since I had to clean up a couple other items as well.